### PR TITLE
Add Grain support for multimodal SFT data processing pipeline

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

This PR migrates the Multimodal Supervised Fine-Tuning (SFT) data pipeline from `dataset_type=hf` to `dataset_type=grain`, with support for single-image (ChartQA), multi-image (SlideVQA), and MRoPE-based models (Qwen3-VL).

### Summary of Changes:
1. **Grain Multimodal SFT Pipeline**:
   - Implemented `vision_sft_preprocessing_pipeline` in `src/maxtext/input_pipeline/grain_data_processing.py` to handle prompt formatting, image decoding, Grain-native tokenization (`grain_tokenizer.TokenizeAndTrim`), prompt masking, batching, image-folding, MRoPE 3D position computation (`ComputeQwen3OmniPositions`), and sequence shifting in Grain.
   - Handled both single-image (`train_image_column: 'image'`) and multi-image columns (`train_image_column: ['page_1', ..., 'page_20']`) via explicit `image_column` parameter.
   - Updated `get_tokenizer_and_pad_id` to allow overriding `add_bos` and `add_eos` to `False` for multimodal SFT pipelines to prevent duplicate control tokens.
   - Updated `_get_pipeline_fn` in `grain_data_processing.py` to route to `vision_sft_preprocessing_pipeline` when `config.use_sft and config.use_multimodal`.
2. **Multimodal Utilities**:
   - Enhanced `convert_to_RGB` in `src/maxtext/multimodal/utils.py` to support dictionary-wrapped bytes (`{'bytes': ...}` as formatted in Parquet/Grain datasets), raw bytes, and file paths.
3. **Configs, Tests & Documentation**:
   - Updated `src/maxtext/configs/post_train/sft-vision-chartqa.yml` and `src/maxtext/configs/post_train/sft-vision-slidevqa.yml` to default to `dataset_type: grain` and `grain_file_type: parquet`.
   - Updated `docs/tutorials/posttraining/multimodal.md` and `tests/end_to_end/tpu/gemma3/4b/test_gemma3_multimodal_sft.sh` to use `dataset_type=grain`.
   - Consolidated unit tests in `tests/unit/grain_sft_test.py` marked with `@pytest.mark.cpu_only` covering ChartQA, SlideVQA multi-image merging, elastic iterator validation, BOS/EOS exclusion, Qwen3-VL MRoPE 3D positions, and exact output batch array equality matching between Hugging Face and Grain pipelines.

# Tests

- **Unit Tests**:
  - `PYTHONPATH=src pytest -m cpu_only tests/unit/grain_sft_test.py` (Passed 6/6 on CPU).
  - `PYTHONPATH=src pytest tests/unit/hf_data_processing_test.py` (Passed 3/3).
  - `PYTHONPATH=src pytest tests/unit/maxtext_utils_test.py -k test_qwen_and_gemma_omit_image_masks` (Passed 1/1).
- **Remote TPU Verification**:
  - **Gemma3-4B SFT**: Executed a 5-step native SFT training run on ChartQA via Grain on TPU v4 (`snehalv-tpu-v4`).
  - **Qwen3-VL-2B SFT**: Executed a native SFT training run with MRoPE positions on ChartQA via Grain on TPU v4 (`snehalv-tpu-v4`) with `model_name=qwen3-vl-2b`, completing training steps and saving checkpoints.
- **Pre-commit Checks**:
  - `pylint` (10.00/10), `pyink` clean.

# Checklist

- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by aireenmei — COMMENTED** [review]
Could you see if we can also migrate https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/configs/post_train/sft-vision-slidevqa.yml with current change? Feel free to leave it to future PR if additional changes needed

File: src/maxtext/input_pipeline/grain_data_processing.py [hunk 1]
```diff
@@ -434,8 +447,143 @@ def sft_preprocessing_pipeline(
...
+      )
+  )
+
+  tokenizer_model, pad_id = data_processing_utils.get_tokenizer_and_pad_id(config)
+  hf_tokenizer = getattr(tokenizer_model, "tokenizer", tokenizer_model)
+
+  if tokenize:
+    dataset = dataset.map(
+        functools.partial(
+            input_pipeline_utils.tokenization,
```
Comment (aireenmei) [inline src/maxtext/input_pipeline/grain_data_processing.py]: `input_pipeline_utils.tokenization` is specific to hf input pipeline, `grain_tokenizer.TokenizeAndTrim` is the counterpart for grain

File: src/maxtext/input_pipeline/grain_data_processing.py [hunk 2]
```diff
@@ -434,8 +447,143 @@ def sft_preprocessing_pipeline(
...
+  """Use grain pipeline to pre-process dataset and return iterators for multimodal SFT fine-tuning."""
+  if config.grain_use_elastic_iterator:
+    raise ValueError(
+        "ElasticIterator is not supported yet for multimodal SFT because post-batch "
+        "transformations (like folding images) cannot be easily applied to the iterator."
+    )
+  assert len(data_columns) == 2, f"Need two data_columns for query and response, received {data_columns=}"
+  text_columns = list(data_columns)
+
+  if list(data_columns) == list(getattr(config, "eval_data_columns", [])):
```
Comment (aireenmei) [inline src/maxtext/input_pipeline/grain_data_processing.py]: Pass image_column as a param similar to data_columns.

File: src/maxtext/input_pipeline/grain_data_processing.py [hunk 3]
```diff
@@ -434,8 +447,143 @@ def sft_preprocessing_pipeline(
...
+
+  dataset = dataset.map(
+      functools.partial(
+          input_pipeline_utils.pre_process_image_sft,
+          image_column="images",
+          config=config,
+      )
+  )
+
+  tokenizer_model, pad_id = data_processing_utils.get_tokenizer_and_pad_id(config)
```
Comment (aireenmei) [inline src/maxtext/input_pipeline/grain_data_processing.py]: The hf vision_sft_pipeline disables bos and eos [here](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/input_pipeline/hf_data_processing.py#L180-L181) regardless of the `add_bos` and `add_eos` flags in base.yml. This is because the reformat prompt and response steps earlier already handles bos/eos. Update get_tokenizer_and_pad_id to allow overriding add_bos and add_eos to false. Also add comments to explain and a unit test make sure no bos, eos in the tokenization output.

File: tests/unit/grain_sft_test.py
Comment (aireenmei) [inline tests/unit/grain_sft_test.py:1]: 1. can we add some tests to make sure the data batch of the two pipelines are matching, not just shape.
2. test_slidevqa_config_initialization seems redundant as sft-vision-slidevqa.yml is already included in BASE_CONFIGS in tests/unit/configs_test.py. 
3. multimodal_sft_grain_test seems very specific to this feature, can we group the test into tests/unit/grain_data_processing_test.py? Or we can make a separate grain_sft_test.py, and we expand it later as we migrate other sft pipelines. Up to you, but try to group the tests properly and leverage existing files and avoid having too many specific unit test files.

Comment (aireenmei) [reply tests/unit/grain_sft_test.py:1]: for 1. tests/post_training/unit/sft_data_processing_test.py is a good example.

**Review by aireenmei — APPROVED** [review]
Thanks for the work! Could you check if the newly added unit tests are running on CPU or TPU? I think they should be fine on CPU since our TPU resource is limited

File: src/maxtext/input_pipeline/input_pipeline_utils.py [hunk 4]
```diff
@@ -315,14 +320,39 @@ def apply_chat_template(example, tokenizer_model, data_column_name):
 
 
 def tokenization(example, hf_tokenizer, truncation, max_length, column_names):
```
Comment (aireenmei) [inline src/maxtext/input_pipeline/input_pipeline_utils.py:354]: Since we uses grain_tokenizer now, I think we don't need to change this function for the feature of this PR

Comment (aireenmei) [inline src/maxtext/input_pipeline/input_pipeline_utils.py:354]: Since we uses grain_tokenizer now, I think we don't need to change this function for the purpose of this PR.

**Review by aireenmei — COMMENTED** [review]
From the PR description, 3. Tokenizer & Worktree Interoperability and 4. IterDataset Conversion seems to be beyond the scope of this PR? Please double check if they are needed to this feature

**Review by aireenmei — COMMENTED** [review]
From the PR description, 3. Tokenizer & Worktree Interoperability and 4. IterDataset Conversion seems to be beyond the scope of this PR? Please double check if they are needed to this feature

**Review by SurbhiJainUSC — APPROVED** [review]
